ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -211,17 +196,85 @@ public void removeImportedTransaction(Xid xid) throws XAException
 	}
 	
 	public Set<Xid> getInflightXids(String parentNodeName) {
-		Iterator<TransactionImple> iterator = _transactions.values().iterator();
+		Iterator<AtomicReference<TransactionImple>> iterator = _transactions.values().iterator();
 		Set<Xid> toReturn = new HashSet<Xid>();
 		while (iterator.hasNext()) {
-			TransactionImple next = iterator.next();
-			if (next.getParentNodeName().equals(parentNodeName)) {
-				toReturn.add(next.baseXid());
+			TransactionImple imported = extractImportedTransaction(iterator.next());
+
+			if (imported != null && imported.getParentNodeName().equals(parentNodeName)) {
+				toReturn.add(imported.baseXid());
 			}
 		}
 		return toReturn;
 	}
 
-	private static ConcurrentHashMap<SubordinateXidImple, TransactionImple> _transactions = new ConcurrentHashMap<SubordinateXidImple, TransactionImple>();
+	private TransactionImple addImportedTransaction(TransactionImple importedTransaction, Xid mapKey, Xid xid, int timeout) {
+		SubordinateXidImple importedXid = new SubordinateXidImple(mapKey);
+		// We need to store the imported transaction in a volatile field holder so that it can be shared between threads
+		AtomicReference<TransactionImple> holder = new AtomicReference<>();
+		AtomicReference<TransactionImple> existing;
+
+		if ((existing = _transactions.putIfAbsent(importedXid, holder)) != null) {
+			holder = existing;
+		}
+
+		TransactionImple txn = holder.get();
+
+		if (txn == null) {
+			// retry the get under a lock - this double check idiom is safe because AtomicReference is effectively
+			// a volatile so can be concurrently accessed by multiple threads
+			synchronized (holder) {
+				txn = holder.get();
+				if (txn == null) {
```

Comment:
Because adding the reference to the map is a distinct atomic action compared to initializing the reference.  If existing is null, all that means is, we were the one to add the reference to the map.  If two threads run at the same time, it is legitimate for the other thread to add the reference and this thread to initialize it; the identity of the reference being added is unimportant as long as it is only added once and all threads recognize the same instance.

What you suggest is possible, but to do it, the reference would have to have been synchronized by the put thread before it is added the map, preventing readers from seeing empty, unlocked reference.  This is also a legitimate solution to the problem, though it might add some complexity to the code due to the strict lexical scoping of the synchronization.

In either case, you could add an optimistic _transactions.get() first, before taking the lock, in order to avoid the expense of the instantiation and (now) lock acquisition if possible (not perfectly of course, but in all but the raciest cases it would yield the optimal path: if the get fails, the putIfAbsent is likely to succeed and thus all the extra work (such as it is) likely to be worthwhile).